code/ecdsap256sha2/Hacl.Impl.ECDSA.P256SHA256.Signature.Agile.fst

```diff
@@ -0,0 +1,319 @@
+      assert_norm (pow2 32 < pow2 61);
+      assert_norm (pow2 32 < pow2 125);
+      let hashM = (hashSpec alg) (as_seq h0 m) in 
+      let cutHashM = Lib.Sequence.sub hashM 0 32 in 
+      as_nat h1 result = nat_from_bytes_be cutHashM % prime_p256_order
+    )
+  )
+
+
+let ecdsa_signature_step12 alg mLen m result = 
+  assert_norm (pow2 32 < pow2 61);
+  assert_norm (pow2 32 < pow2 125);
+  push_frame(); 
+    let h0 = ST.get() in 
+  let sz: FStar.UInt32.t = hash_len alg in
+  let mHash = create sz (u8 0) in    
+  
+  begin
+  match alg with 
+    |SHA2_256 ->
+      hash_256 m mLen mHash
+    |SHA2_384 ->
+      hash_384 m mLen mHash
+    |SHA2_512 -> 
+      hash_512 m mLen mHash
+  end;
+  
+  let cutHash = sub mHash (size 0) (size 32) in 
+  toUint64ChangeEndian cutHash result;
+  
+  let h1 = ST.get() in 
+ 
+  reduction_prime_2prime_order result result;
+
+  lemma_core_0 result h1;
+  Spec.ECDSA.changeEndianLemma (uints_from_bytes_be #U64 #_ #4 (as_seq h1 cutHash));
+  uints_from_bytes_be_nat_lemma #U64 #_ #4 (as_seq h1 cutHash);
+
+  pop_frame()
+
+
+inline_for_extraction
+val ecdsa_signature_step45: x: felem -> k: lbuffer uint8 (size 32) -> tempBuffer: lbuffer uint64 (size 100) -> Stack uint64
+  (requires fun h -> 
+    live h x /\ live h k /\ live h tempBuffer /\ 
+    LowStar.Monotonic.Buffer.all_disjoint [loc tempBuffer; loc k; loc x]
+  )
+  (ensures fun h0 r h1 -> 
+    modifies (loc x |+| loc tempBuffer) h0 h1 /\ 
+    as_nat h1 x < prime_p256_order /\ 
+    (
+      let (rxN, ryN, rzN), _ = montgomery_ladder_spec (as_seq h0 k) ((0,0,0), basePoint) in 
+      let (xN, _, _) = _norm (rxN, ryN, rzN) in 
+      as_nat h1 x == xN % prime_p256_order /\ 
+      (
+  if as_nat h1 x = 0 then uint_v r == pow2 64 - 1 else uint_v r == 0
+      )
+    )
+  )
+
+let ecdsa_signature_step45 x k tempBuffer = 
+  push_frame();
+    let result = create (size 12) (u64 0) in 
+    let tempForNorm = sub tempBuffer (size 0) (size 88) in 
+    secretToPublicWithoutNorm result k tempBuffer; 
+    normX result x tempForNorm;
+    reduction_prime_2prime_order x x;
+  pop_frame();
+    isZero_uint64_CT x
+
+
+val lemma_power_step6: kInv: nat -> Lemma 
+  (Spec.ECDSA.exponent_spec (fromDomain_ kInv) == toDomain_ (pow kInv (prime_p256_order - 2)))
+
+let lemma_power_step6 kInv = 
+  let a = Spec.ECDSA.exponent_spec (fromDomain_ kInv) in 
+  lemmaFromDomain kInv;
+
+  power_distributivity (kInv * modp_inv2_prime (pow2 256) prime_p256_order) (prime_p256_order - 2) prime_p256_order;
+  power_distributivity_2 kInv (modp_inv2_prime (pow2 256) prime_p256_order % prime_p256_order) (prime_p256_order - 2);
+  lemma_mod_mul_distr_r (pow kInv (prime_p256_order - 2)) (pow (modp_inv2_prime (pow2 256) prime_p256_order) (prime_p256_order - 2)) prime_p256_order;
+
+  lemma_pow_mod_n_is_fpow prime_p256_order (pow2 256 % prime_p256_order) (prime_p256_order - 2);
+  
+  let inverse2_256 = 43790243014242295660885426880012836369732278457577312309071968676491870960761 in 
```

Comment:
one good thing for readability would be to try to avoid magic constants -- for instance, here, you could do:

```
let inverse2_256 = normalize_expr (modp_inv2_prime (pow2 256) prime_p256_order) in
FStar.Pervasives.normalize_term_spec inverse2_256 (modp_inv2_prime (pow2 256) prime_p256_order);
```